### ISSUE TITLE
Representing transcript information to predict future course success

### ISSUE THREAD

**Issue by droober** [rstudio/keras#820]
I am working with a large amount of student transcript information and trying to use Keras in R to predict future course success.  I am new to deep learning but have just finished reading _Deep Learning with R_ and am trying to apply what i've learned to this project.  I am not sure if this is the right place to post; if not please just close/delete this issue. 

I can derive many very predictive numerical features from transcripts (such as GPA) and use techniques like extreme gradient boosting to generate some very successful models. But I want to explore the possibility of using Deep Learning to take some of text data from a transcript and see if it has any predictive power.

When calculating GPA I ignore the specific combination of courses that a student has taken to get that GPA. For example, two students taking English 200 with a 3.0 GPA should not possess the exact level of preparation because they've probably taken different courses. My goal is to feed a model all the grades and courses a student has taken plus the course they are currently attempting to see if it can discover types of preparation that yield higher pass rates than others.

I have tried a couple of different techniques with mediocre results.

1) Represent the data sparsely and use a fully connected network. Each row of data contains 200 columns for the 200 courses possible courses they may be currently attempted. Then another 1000 columns as flags for each possible course/grade on their transcript. Giving this data to a fully connected network with a couple of hidden layers gets me a test data accuracy of about 68% (for a baseline I'm using the average pass rate in the test data of 61%). As expected, this model performs better and better on the training data but quickly stops improving on validation data.

2) Represent the data as a list of words and use and embedding layer followed by a 1D convnet. Each row of data is 
1) the course they are attempting 
2) course 1 on their transcript
3) grade in course 1
4) course 2 on their transcript
5) grade in course 2
.
.
20) course 10 on their transcript
21) grade in course 10.

If the student has more than 10 courses on their transcript i cut it off taking the most recent 10. I am coding the course attempted field with a different set of values than when the course may appear in their transcript history. I am only just starting to wrap my head around how 1D convnets work and am not very sure that I am setting up my model in a way that makes sense. Right now i have it set up like this:

```
model <- keras_model_sequential() %>%
   layer_embedding(input_dim = 2123, output_dim = 128, input_length = 22) %>%
   layer_conv_1d(filters = 128, kernel_size = 2, activation = "relu", strides = 2) %>%
   layer_max_pooling_1d(pool_size = 2, strides = 2) %>%
   layer_conv_1d(filters = 128, kernel_size = 2, activation = "relu", strides = 2) %>%
   layer_global_average_pooling_1d() %>%
   layer_dropout(0.2) %>%
   layer_dense(units = 1, activation="sigmoid")
```

This set up performs similarly with test data accuracy about 6-7 percentage points higher than my test data baseline. I suspect I am doing something wrong here because the training process jumps all over the place and my accuracy on training data nor validation data systematically improves over the epochs of its training.  I am also leaving the timing of their course history out of this model mainly because i do not know how to incorporate it.

Neither performance is really much better than if I just calculate a students GPA and throw that into randomforest or a logistic regression. This seems to indicate that the network is not doing anything valuable with the extra transcript data I am giving it. At best it is simply engineering something similar to GPA. Any suggestions for how I might differently represent or model this data would be greatly appreciated. Thanks!

**Comment by skeydan** [comment]
Hi,

I'm not sure I completely understand the task (a bit overwhelmed by the amount of information :-)).
Is it ok if I abstract to: _how to combine numerical data and textual data_? Equivalently, how to enrich numerical data with text data that happen to supplement it?

If that matches approximately, you could take a look at https://blogs.rstudio.com/tensorflow/posts/2018-11-26-embeddings-fun-and-profit/, which shows a few possible ways. If you have more text than in that example, you might follow the same strategy, but incorporating a convnet or an lstm, that is

- process numerical data with dense layers
- process text with embeddings, then try convnet or lstm, then flatten
- concatenate 

However, how much that improves performance over a numeric-data model only isn't possible to know in advance...

**Comment by droober** [comment]
Thank you for your comment.  I will take a look at this blog.
